Modules/Core/Common/include/itkBSplineInterpolationWeightFunction.h

```diff
@@ -112,15 +113,21 @@ class ITK_TEMPLATE_EXPORT BSplineInterpolationWeightFunction
 #endif
 
 protected:
-  BSplineInterpolationWeightFunction();
+  BSplineInterpolationWeightFunction() = default;
   ~BSplineInterpolationWeightFunction() override = default;
 
 private:
   /** Lookup table type. */
-  using TableType = Array2D<unsigned int>;
+  using TableType = FixedArray<IndexType, NumberOfWeights>;
 
   /** Table mapping linear offset to indices. */
-  TableType m_OffsetToIndexTable;
+  const TableType m_OffsetToIndexTable{ [] {
+    TableType     table;
+    // Note: Copied the constexpr value `SupportSize` to a temporary, `SizeType{ SupportSize }`, to prevent a GCC
```

Comment:
Good solution!

Reply:
@dzenanz Thanks! It appears that GCC doesn't like to have a const-reference to a static `constexpr` data member. The syntax `T{x}` produces a copy that may be passed to a const-reference parameter, so that appears to solve the errors. (The copying itself is easily optimized by the compiler, I assume.) 

Note that this issue may appear more often with a constexpr static `itk::Size<N>` data member than with constexpr static data members of a built-in type, because `itk::Size<N>` is more likely to be passed to a const-reference parameter, while built-in types are more likely passed by-value. (Pass-by-value is not a problem, as far as I can see.)

Reply:
> It appears that GCC doesn't like to have a const-reference to a static constexpr data member. 

This can happen when the static constexpr member is declared within the class declaration but not defined outside of the class declaration, and then a reference is used.  In the associated hxx file define the member -- omit the `static` keyword in the hxx file and don't initialize the variable if it is already initialized within the class declaration.

Reply:
> In the associated hxx file define the member -- omit the `static` keyword in the hxx file and don't initialize the variable if it is already initialized within the class declaration.

Doing so might solve the problem, thanks @Leengit However, I feel that it's unfortunate if it is necessary to do it that way. The following is accepted, even by GCC 7.5: https://godbolt.org/z/zb79f7v1E as below:

```
int square(const long long& ref)
{
    return static_cast<int>(ref*ref);
}

int main()
{
    constexpr int i{42};
    return square(i);
}
```

You see, the `ref` parameter refers to a `long long` that isn't explicitly declared or defined anywhere! So do you think it's a compiler issue, or a language issue, that we need to work around those "undefined reference" errors?

Reply:
I don't claim to know what the GCC folks are thinking nor what the C++ language spec says on the matter, but I've always imagined that the thinking is something like the following.  The implementation of the likes of
```c++
static const int CrucialParameter = 3;
```
(or `constexpr` instead of `const`) within a class declaration is usually never allocated in memory.  Usually the compiler treats it much as
```c++
#define CrucialParameter 3
```
would be treated (though with class namespace scoping).  That works great unless one wants a pointer or reference to `CrucialParameter`.  The way to tell GCC about this added need is define the member too, sometime after the class declaration is closed:
```
const inst MyClass::CrucialParameter;
```

The code sample you supply doesn't includes a `static const` so I'm not sure that it applies to the present case.

Reply:
> That works great unless one wants a pointer or reference to `CrucialParameter`. The way to tell GCC about this added need is define the member too, sometime after the class declaration is closed:

@Leengit OK, but then, do you think ideally all "static constexpr" data should have a separate definition, apart from its initialization, in ITK? A quick search suggests that ITK now declares more than 2000 "static constexpr" constants already! (Quite some of them from ThirdParty, I have to admit.)

Reply:
I suppose if we think it will happen often enough that a user of ITK will want a pointer or reference to `CrucialParameter` then we might want to proactively define it outside the class declaration.  I have no opinion on that possibility.

Except for that circumstance (and except for ITK's direct use of a pointer or reference), the optimization of not allocating memory is the desired outcome; we would deliberately not add the definition outside of the class declaration.